File: 23_设计类/设计版本控制(可持久化)/可持久化数组/PersistentQueue/RealTimeQueue.go
// https://scrapbox.io/data-structures/Realtime_Queue

package main

import (
	"bufio"
	"fmt"
	"os"
	"runtime/debug"
	"time"
)

func init() {
	debug.SetGCPercent(-1)
}

func main2() {
	assert := func(a bool) {
		if !a {
			panic("assert fail")
		}
	}
	queue := NewRealTimeQueue()
	assert(queue.Empty() == true)
	queue1 := queue.Push(1)
	assert(queue1.Empty() == false)
	assert(queue1.Front() == 1)
	queue2 := queue1.Push(2)
	assert(queue2.Empty() == false)
	assert(queue2.Front() == 1)
	queue3 := queue2.Shift()
	assert(queue3.Empty() == false)
	assert(queue3.Front() == 2)
	queue4 := queue3.Shift()
	assert(queue4.Empty() == true)

	time1 := time.Now()
	for i := 0; i < 1e6; i++ {
		queue = queue.Push(i)
		queue = queue.Shift()
		queue.Empty()
	}

	fmt.Println(time.Since(time1)) // 343ms
}

func main() {
	// https://judge.yosupo.jp/problem/persistent_queue
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var q int
	fmt.Fscan(in, &q)

	gits := make([]*RealTimeQueue, q+1)
	gits[0] = NewRealTimeQueue()

	for i := 0; i < q; i++ {
		var op, time, x int
		fmt.Fscan(in, &op, &time)
		time++
		if op == 0 {
			fmt.Fscan(in, &x)
			gits[i+1] = gits[time].Push(x)
		} else {
			res := gits[time].Front()
			fmt.Fprintln(out, res)
			gits[i+1] = gits[time].Shift()
		}
	}
}

type S = int

// 可持久化队列.
//  https://scrapbox.io/data-structures/Realtime_Queue
type RealTimeQueue struct {
	front    *_Stream
	back     *_PersistentStack
	schedule *_Stream
}

func NewRealTimeQueue() *RealTimeQueue {
	return &RealTimeQueue{
		front:    _NewStream(),
		back:     _NewPersistentStack(),
		schedule: _NewStream(),
	}
}

func (q *RealTimeQueue) Empty() bool {
	return q.front.Empty()
}

func (q *RealTimeQueue) Front() S {
	if q.Empty() {
		panic("queue is empty")
	}
	return q.front.Top()
}

func (q *RealTimeQueue) Push(x S) *RealTimeQueue {
	return q._makeQueue(q.front, q.back.Push(x), q.schedule)
}

func (q *RealTimeQueue) Shift() *RealTimeQueue {
	if q.Empty() {
		return NewRealTimeQueue()
	}
	return q._makeQueue(q.front.Pop(), q.back, q.schedule)
}

func (q *RealTimeQueue) _rotate(f *_Stream, b *_PersistentStack, s *_Stream) *_Stream {
	return &_Stream{_Suspension: _NewSuspensionWith(func() interface{} {
		if f.Empty() {
			return &_Cell{resolved: b.Top(), next: s}
		}
		return &_Cell{resolved: f.Top(), next: q._rotate(f.Pop(), b.Pop(), s.Push(b.Top()))}
	})}
}

func (q *RealTimeQueue) _makeQueue(f *_Stream, b *_PersistentStack, s *_Stream) *RealTimeQueue {
	if !s.Empty() {
		return &RealTimeQueue{front: f, back: b, schedule: s.Pop()}
	}
	tmp := q._rotate(f, b, _NewStream())
	return &RealTimeQueue{front: tmp, back: _NewPersistentStack(), schedule: tmp}
}

type _Cell struct {
	resolved S
	next     *_Stream
}

type _Stream struct {
	*_Suspension
}

// 惰性求值的流.
func _NewStream() *_Stream {
	return &_Stream{_Suspension: _NewSuspension()}
}

func (s *_Stream) Empty() bool {
	return s.Resolve() == nil
}

func (s *_Stream) Top() S {
	return s.Resolve().(*_Cell).resolved
}

func (s *_Stream) Pop() *_Stream {
	return s.Resolve().(*_Cell).next
}

func (s *_Stream) Push(x S) *_Stream {
	return &_Stream{_Suspension: _NewSuspensionWith(&_Cell{resolved: x, next: s})}
}

func (s *_Stream) String() string {
	x := s
	res := []S{}
	for !x.Empty() {
		res = append(res, x.Top())
		x = x.Pop()
	}
	for i, j := 0, len(res)-1; i < j; i, j = i+1, j-1 {
		res[i], res[j] = res[j], res[i]
	}
	return fmt.Sprintf("Stream%v", res)
}

// 惰性求值.
type _Suspension struct {
	x        interface{}
	resolved interface{}
}

func _NewSuspension() *_Suspension {
	return &_Suspension{}
}

func _NewSuspensionWith(x interface{}) *_Suspension {
	return &_Suspension{x: x}
}

func (s *_Suspension) Resolve() interface{} {
	if s.resolved == nil {
		if f, ok := s.x.(func() interface{}); ok {
			s.resolved = f()
		} else {
			s.resolved = s.x
		}
	}
	return s.resolved
}

type _PersistentStack struct {
	root *_StackNode
}

type _StackNode struct {
	value S
	pre   *_StackNode
}

// 创建一个新的可持久化栈.
func _NewPersistentStack() *_PersistentStack {
	return &_PersistentStack{}
}

func (s *_PersistentStack) Push(value S) *_PersistentStack {
	return &_PersistentStack{root: &_StackNode{value: value, pre: s.root}}
}

func (s *_PersistentStack) Pop() *_PersistentStack {
	if s.root == nil {
		panic("stack is empty")
	}
	return &_PersistentStack{root: s.root.pre}
}

func (s *_PersistentStack) Top() S {
	if s.root == nil {
		panic("stack is empty")
	}
	return s.root.value
}

func (s *_PersistentStack) Empty() bool {
	return s.root == nil
}

func (s *_PersistentStack) Reverse() *_PersistentStack {
	res := _NewPersistentStack()
	x := s
	for !x.Empty() {
		res = res.Push(x.Top())
		x = x.Pop()
	}
	return res
}

func (s *_PersistentStack) String() string {
	sb := []string{}
	x := s
	for !x.Empty() {
		sb = append(sb, fmt.Sprintf("%v", x.Top()))
		x = x.Pop()
	}
	for i, j := 0, len(sb)-1; i < j; i, j = i+1, j-1 {
		sb[i], sb[j] = sb[j], sb[i]
	}
	return fmt.Sprintf("Stack%v", sb)
}

File: 23_设计类/设计版本控制-可持久化/可持久化数组/PersistentQueue/RealTimeQueue.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */
// https://scrapbox.io/data-structures/Realtime_Queue

import assert from 'assert'

import { Stream } from './Stream'
import { PersistentStack } from '../../可持久化栈/PersistentStack'

/**
 * 可持久化队列.
 * @link https://scrapbox.io/data-structures/Realtime_Queue
 */
class RealTimeQueue<T> {
  private static _rotate<T>(f: Stream<T>, b: PersistentStack<T>, t: Stream<T>): Stream<T> {
    return new Stream(() => {
      if (f.empty()) {
        return [b.top()!, t]
      }
      return [f.top()!, this._rotate(f.pop()!, b.pop()!, t.push(b.top()!))]
    })
  }

  private static _makeQueue<T>(f: Stream<T>, b: PersistentStack<T>, s: Stream<T>): RealTimeQueue<T> {
    if (!s.empty()) {
      return new RealTimeQueue(f, b, s.pop()!)
    }
    const tmp = this._rotate(f, b, new Stream())
    return new RealTimeQueue(tmp, new PersistentStack(), tmp)
  }

  private readonly _front = new Stream<T>()
  private readonly _back = new PersistentStack<T>()
  private readonly _schedule = new Stream<T>()

  constructor()
  constructor(f: Stream<T>, b: PersistentStack<T>, s: Stream<T>)
  constructor(f?: Stream<T>, b?: PersistentStack<T>, s?: Stream<T>) {
    const hasArgs = !!arguments.length // 实参数量
    if (hasArgs) {
      this._front = f!
      this._back = b!
      this._schedule = s!
    }
  }

  empty(): boolean {
    return this._front.empty()
  }

  front(): T | undefined {
    return this._front.top()
  }

  push(x: T): RealTimeQueue<T> {
    return RealTimeQueue._makeQueue(this._front!, this._back!.push(x), this._schedule!)
  }

  shift(): RealTimeQueue<T> {
    if (this.empty()) {
      return new RealTimeQueue()
    }
    return RealTimeQueue._makeQueue(this._front!.pop()!, this._back!, this._schedule!)
  }
}

export { RealTimeQueue, RealTimeQueue as PersistentQueue }

if (require.main === module) {
  let queue = new RealTimeQueue<number>()

  assert(queue.empty())
  assert(queue.front() === undefined)
  const queue1 = queue.push(1)
  assert(!queue1.empty())
  assert(queue1.front() === 1)

  const queue2 = queue1.push(2)
  assert(!queue2.empty())
  assert(queue2.front() === 1)

  const queue3 = queue2.shift()
  assert(!queue3.empty())
  assert(queue3.front() === 2)

  const queue4 = queue3.shift()
  assert(queue4.empty())
  assert(queue4.front() === undefined)

  console.time('queue')
  for (let i = 0; i < 1e7; i++) {
    queue = queue.push(i)
    queue = queue.shift()
    queue.empty()
    queue.front()
  }
  console.timeEnd('queue') // queue: 1.447s
}
